bftengine/src/bftengine/KeyManager.cpp

```diff
@@ -0,0 +1,95 @@
+#include "KeyManager.h"
+#include "thread"
+#include "ReplicaImp.hpp"
+#include "ReplicaConfig.hpp"
+
+// KeyManager::KeyExchangeMsg::KeyExchangeMsg(KeyManager::KeyExchangeMsg&&
+// other):key_(std::move(other.key_)),signature_(other.signature_),repID_(other.repID_){}
+
+const std::string KeyManager::KeyExchangeMsg::getVersion() const { return "1"; }
+
+KeyManager::KeyExchangeMsg KeyManager::KeyExchangeMsg::KeyExDes(const char* serStr, const uint32_t& size) {
+  std::stringstream ss;
+  KeyManager::KeyExchangeMsg ke;
+  ss.write(serStr, std::streamsize(size));
+  KeyManager::KeyExchangeMsg::deserialize(ss, ke);
+  return ke;
+}
+
+void KeyManager::KeyExchangeMsg::serializeDataMembers(std::ostream& outStream) const {
+  serialize(outStream, key_);
+  LOG_TRACE(GL, "KEY EXCHANGE MANAGER  ser  key_ " << key_);
+  serialize(outStream, signature_);
+  LOG_TRACE(GL, "KEY EXCHANGE MANAGER  ser  signature_ " << signature_);
+  serialize(outStream, repID_);
+  LOG_TRACE(GL, "KEY EXCHANGE MANAGER  ser  repID_ " << repID_);
+}
+
+void KeyManager::KeyExchangeMsg::deserializeDataMembers(std::istream& inStream) {
+  deserialize(inStream, key_);
+  deserialize(inStream, signature_);
+  deserialize(inStream, repID_);
+}
+
+std::string KeyManager::KeyExchangeMsg::toString() const {
+  std::stringstream ss;
+  ss << "key [" << key_ << "] signature [" << signature_ << "] replica id [" << repID_ << "]";
+  return ss.str();
+}
+
+KeyManager& KeyManager::get() {
+  static KeyManager km;
+  return km;
+}
+void KeyManager::set(concord::kvbc::IClient* cl, const int& id) {
+  cl_ = cl;
+  repID_ = id;
+}
+
+void KeyManager::start(const int& durMs) {
+  auto l = [this](const int& durMs) {
+    while (true) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(durMs));
+      sendKeyExchange();
+    }
+  };
+  std::thread t(l, durMs);
+  t.detach();
```

Comment:
Detaching threads is highly not recommended:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconc-detached_thread

Reply:
It's a test function, will be deleted once real generation takes place